### PR TITLE
Caps type attrs

### PR DESCRIPTION
**Pull request by zyga**
This branch adds simple attribute validation to capability types. Each type now contains the list of attributes that must (and only those) be present in a capability.

### CODE REVIEW

File: caps/types.go [hunk 1]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
 	if t != c.Type {
 		return fmt.Errorf("capability is not of type %q", t)
 	}
-	// While we don't have any support for type-specific attribute schema,
-	// let's ensure that attributes are totally empty. This will make tests
-	// show that this code is actually being used.
-	if c.Attrs != nil && len(c.Attrs) != 0 {
-		return fmt.Errorf("attributes must be empty for now")
+	// Check that all supported attributes are present
+	for _, attr := range t.SupportedAttrs {
```
Comment (mvo5) [inline caps/types.go]: Hm, if I read this correctly it will fail if a "SupportedAttr" is missing? If thats the case, maybe it should be "t.RequiredAttrs" instead as the name? To make more clear that this is a mandatory attribute?

Comment (zyga) [inline caps/types.go]: I was actually looking for "mandatory/required + only (no other allowed).

I'll rename this to RequiredAttrs for now

File: caps/types.go [hunk 2]
```diff
@@ -31,6 +31,9 @@ type Type struct {
 	// Name is a key that identifies the capability type. It must be unique
 	// within the whole OS. The name forms a part of the stable system API.
 	Name string
+	// RequiredAttrs contains names of attributes that are understood by
```
Comment (mvo5) [inline caps/types.go]: In nitpick mode now (sorry for that!). Maybe instead of "that are understood by" "that are required by" ?

File: caps/types.go [hunk 3]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
 	if t != c.Type {
 		return fmt.Errorf("capability is not of type %q", t)
 	}
-	// While we don't have any support for type-specific attribute schema,
-	// let's ensure that attributes are totally empty. This will make tests
-	// show that this code is actually being used.
-	if c.Attrs != nil && len(c.Attrs) != 0 {
-		return fmt.Errorf("attributes must be empty for now")
+	// Check that all supported attributes are present
```
Comment (mvo5) [inline caps/types.go]: (nitpick) supported -> required (?)

File: caps/types.go (hunk 2 above)
Comment (zyga) [inline caps/types.go]: Ack, corrected.

File: caps/types.go [hunk 4]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
...
+	for _, attr := range t.RequiredAttrs {
+		if _, ok := c.Attrs[attr]; !ok {
+			return fmt.Errorf("capability lacks required attribute %q", attr)
+		}
+	}
+	// Look for any unexpected attributes
+	if len(t.RequiredAttrs) != len(c.Attrs) {
+		for attr := range c.Attrs {
+			supported := false
```
Comment (mvo5) [inline caps/types.go]: (nitpick, probably over the top this time). Its a bit unfortunate that there is no "Contains()" outside of the testutil package :) I think this would be easier to read by:
```
for attr := range c.Attrs {
			if !Contains(t.RequiredAttrs, attr)  {
				return fmt.Errorf("capability contains unexpected attribute %q", attr)
			}
```
But not for this branch :)

File: caps/types.go (hunk 3 above)
Comment (zyga) [inline caps/types.go]: Fixed

**Comment by mvo5** [comment]
Thanks, looks good to me.

File: caps/types.go [hunk 5]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
...
+			for _, attrRequired := range t.RequiredAttrs {
+				if attr == attrRequired {
+					supported = true
+					break
+				}
+			}
+			if !supported {
+				return fmt.Errorf("capability contains unexpected attribute %q", attr)
```
Comment (chipaca) [inline caps/types.go]: I think you should wait until the end of the loop and report *all* unexpected attribtues, not just the first one at random.

Comment (chipaca) [inline caps/types.go]: you've got to be <---------- this ----------> nitpicky to follow this suggestion

**Comment by chipaca** [comment]
lgtm

File: caps/types.go [hunk 6]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
...
 	}
-	// While we don't have any support for type-specific attribute schema,
-	// let's ensure that attributes are totally empty. This will make tests
-	// show that this code is actually being used.
-	if c.Attrs != nil && len(c.Attrs) != 0 {
-		return fmt.Errorf("attributes must be empty for now")
+	// Check that all required attributes are present
+	for _, attr := range t.RequiredAttrs {
+		if _, ok := c.Attrs[attr]; !ok {
+			return fmt.Errorf("capability lacks required attribute %q", attr)
```
Comment (niemeyer) [inline caps/types.go]: ("capabilities of type %q must provide a %q attribute", t, attr)

File: caps/types.go [hunk 7]
```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
...
+		}
+	}
+	// Look for any unexpected attributes
```
Comment (niemeyer) [inline caps/types.go]: This will break forwards compatibility in a really bad way. Think in terms of code: imagine you have a Python type that millions of people are using, and then when you decide to add a new method on that type, you break every single existing application out there.

File: caps/types_test.go [hunk 8]
```diff
@@ -50,17 +53,34 @@ func (s *TypeSuite) TestValidateOK(c *C) {
...
+func (s *TypeSuite) TestValidateAttributesRequiredAttrs(c *C) {
+	t := &Type{
+		Name:          "t",
+		RequiredAttrs: []string{"k"},
+	}
+	cap := &Capability{
+		Name:  "name",
+		Label: "label",
+		Type:  t,
+		Attrs: map[string]string{},
```
Comment (niemeyer) [inline caps/types_test.go]: Why do we need an empty map here? The logic should handle nil maps okay, right?

There should also be another test with a non-empty map but non-matching attributes, and also another positive test which has both matching and non-matching attributes but the required attributes satisfied.

File: caps/types.go (hunk 6 above)
Comment (zyga) [inline caps/types.go]: Fixed

File: caps/types.go (hunk 7 above)
Comment (zyga) [inline caps/types.go]: Can you explain this more please? I've made it precisely so that we have forward compatibility. Otherwise anyone can ship a product with a "foo" type that has a custom, unexpected "bar" attribute that we want to use but now cannot because shipping products and snaps have already used it.

Down the line I have an idea for type versions so that we can release, say, "bool-file" v2 with more attributes and applications that want to see that can use the new attributes (perhaps with new usage semantics as well, that are not strictly encoded in attributes but this is unrelated).

When we ship a type we ship it with fixed set of attributes. Each time we want to change what they are, what they mean, we bump the version and ship another capability (version) that represents the same resource. This gives us future-compatibility by letting us evolve types over time.

Obviously you had something else on your mind so please explain what you mean again and I'll adjust the code.

File: caps/types_test.go (hunk 8 above)
Comment (zyga) [inline caps/types_test.go]: No need, removed now.

The second part is fixed with the exception that the last test you've mentioned is not implemented (since I didn't change Validate() yet as I need to wait for your explanation).

**Comment by niemeyer** [comment]
Just for record keeping, this still has some issues to address.

**Comment by zyga** [comment]
I think this is fixed now

**Comment by chipaca** [comment]
@niemeyer AFAICT everything was addressed; landing this now (if we missed anything, there's plenty of branches in the pipeline where we can address whatever it was).
